kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java

```diff
@@ -932,7 +934,7 @@ private void publishMessages(final Optional<PersistentTopic> persistentTopicOpt,
         final Producer producer = KafkaTopicManager.getReferenceProducer(partitionName);
         producer.updateRates(numMessages, byteBuf.readableBytes());
         producer.getTopic().incrementPublishCount(numMessages, byteBuf.readableBytes());
-        updateProducerStats(topicPartition, numMessages, byteBuf.readableBytes());
+        updateProducerStats(topicPartition, numMessages, byteBuf.readableBytes(), conversionCount);
```

Comment:
We can modify the method signature in case that we may add more fields to `EncodeResult`. For example,

```java
    private void updateProducerStats(final TopicPartition topicPartition,
                                     final EncodeResult encodeResult) {
        final int numMessages = encodeResult.getNumMessages();
        final int numBytes = encodeResult.getEncodedByteBuf().readableBytes();

        final Producer producer = KafkaTopicManager.getReferenceProducer(KopTopic.toString(topicPartition));
        producer.updateRates(numMessages, numBytes);
        producer.getTopic().incrementPublishCount(numMessages, numBytes);

        // TODO: update requestStats
    }
```
```

Reply:
Or add this method to `EncodeResult` itself.

```java
    public void updateStats(final TopicPartition topicPartition,
                            final StatsLogger statsLogger) {
        final int numBytes = encodedByteBuf.readableBytes();

        final Producer producer = KafkaTopicManager.getReferenceProducer(KopTopic.toString(topicPartition));
        producer.updateRates(numMessages, numBytes);
        producer.getTopic().incrementPublishCount(numMessages, numBytes);

        statsLogger.scopeLabel(TOPIC_SCOPE, topicPartition.topic())
                .scopeLabel(PARTITION_SCOPE, String.valueOf(topicPartition.partition()))
                .getCounter(BYTES_IN)
                .add(numBytes);

        statsLogger.scopeLabel(TOPIC_SCOPE, topicPartition.topic())
                .scopeLabel(PARTITION_SCOPE, String.valueOf(topicPartition.partition()))
                .getCounter(MESSAGE_IN)
                .add(numMessages);

        if (conversionCount > 0) {
            statsLogger.scopeLabel(TOPIC_SCOPE, topicPartition.topic())
                    .getCounter(PRODUCE_MESSAGE_CONVERSIONS)
                    .add(conversionCount);
        }

        RequestStats.BATCH_COUNT_PER_MEMORY_RECORDS_INSTANCE.set(numMessages);
    }
```

BTW, I'm not sure whether

```java
        statsLogger.scopeLabel(TOPIC_SCOPE, topicPartition.topic())
                .scopeLabel(PARTITION_SCOPE, String.valueOf((topicPartition.partition())))
```

can be reused like

```java
        final StatsLogger statsLoggerForThisPartition = statsLogger
                .scopeLabel(TOPIC_SCOPE, topicPartition.topic())
                .scopeLabel(PARTITION_SCOPE, String.valueOf(topicPartition.partition()));

        statsLoggerForThisPartition.getCounter(BYTES_IN).add(numBytes);
        statsLoggerForThisPartition.getCounter(MESSAGE_IN).add(numMessages);
        statsLoggerForThisPartition.getCounter(PRODUCE_MESSAGE_CONVERSIONS).add(conversionCount);
```

Reply:
Good suggestion. I think there is no problem with the same `StatsLogger` instance. Addressed comments. PTAL again. Thanks.

Reply:
Yeah, I also double checked it yesterday.